// micro-form/dist/lib/withFormState.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _microstate = require('microstate');

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var keys = function keys(o) {
  return Object.keys(o);
};

var withFormState = function withFormState(Component) {
  return (0, _microstate.connect)(function (state) {
    return {
      state: state,
      valid: keys(state).filter(function (key) {
        return state[key].valid === false;
      }).length < 1,
      getPayload: function getPayload() {
        return keys(state).reduce(function (payload, k) {
          payload[k] = state[k].value;
          return payload;
        }, {});
      }
    };
  }, function (dispatch, state, initialState) {
    return {
      resetForm: function resetForm() {
        return dispatch(initialState);
      },
      validateForm: function validateForm(cb) {
        return keys(state).reduce(function (res, key) {
          var valid = state[key].validate(state[key].value);

          if (!valid) {
            res = false;
          }

          dispatch(function (state) {
            return _defineProperty({}, key, {
              value: state[key].value,
              valid: valid,
              validate: state[key].validate
            });
          });

          return res;
        }, true);
      }
    };
  })(Component);
};

exports.default = withFormState;
